packages/caravan-fees/src/cpfp.ts

```diff
@@ -0,0 +1,464 @@
+      ),
+    };
+  }
+
+  /**
+   * Calculates the total input value of the child transaction.
+   *
+   * This method sums up the values of all inputs in the child transaction.
+   * It's used to determine if additional inputs are needed and to calculate fees.
+   *
+   * @private
+   * @returns {BigNumber} The total input value
+   */
+  private calculateTotalInputValue(): BigNumber {
+    return this.childPsbt.PSBT_IN_WITNESS_UTXO.reduce(
+      (sum, utxo) => sum.plus(utxo ? new BigNumber(utxo.split(",")[0]) : 0),
+      new BigNumber(0),
+    );
+  }
+
+  /**
+   * Calculates the required amount for the child transaction.
+   *
+   * This method determines the total amount needed to cover both the desired
+   * output and the estimated fee. It's used to check if additional inputs are required.
+   *
+   * @private
+   * @returns {BigNumber} The required amount
+   */
+  private calculateRequiredAmount(): BigNumber {
+    const outputValue = new BigNumber(
+      this.childPsbt.PSBT_OUT_AMOUNT[0].toString(),
+    );
+    const estimatedFee = this.estimateRequiredFee();
+    return outputValue.plus(estimatedFee);
+  }
+
+  /**
+   * Estimates the required fee for the child transaction.
+   *
+   * This method calculates the estimated fee based on the virtual size of the
+   * transaction and the new fee rate. It's crucial for determining the appropriate
+   * fee to ensure the CPFP transaction is effective.
+   *
+   * @private
+   * @returns {BigNumber} The estimated required fee
+   */
+  private estimateRequiredFee(): BigNumber {
+    const vsize = this.estimateVsize();
+    return new BigNumber(vsize).multipliedBy(
+      this.calculateNewFeeRate().satoshisPerByte,
+    );
+  }
+
+  /**
+   * Calculates the current fee of the child transaction.
+   *
+   * This method computes the difference between the total input value and the
+   * output value, which represents the current fee. It's used to compare against
+   * the new fee when adjusting the fee rate.
+   *
+   * @private
+   * @returns {BigNumber} The current fee
+   */
+  private calculateCurrentFee(): BigNumber {
+    const inputValue = this.calculateTotalInputValue();
+    const outputValue = new BigNumber(
+      this.childPsbt.PSBT_OUT_AMOUNT[0].toString(),
+    );
+    return inputValue.minus(outputValue);
+  }
+
+  private estimateVsize(): number {
+    let totalWeight = 0;
+    // Base transaction overhead
+    totalWeight += 10 * 4; // version, locktime, input count, output count
+    // Calculate weight for inputs
+    for (let i = 0; i < this.childPsbt.PSBT_GLOBAL_INPUT_COUNT; i++) {
+      totalWeight += this.estimateInputWeight(i);
+    }
+    // Calculate weight for outputs
+    for (let i = 0; i < this.childPsbt.PSBT_GLOBAL_OUTPUT_COUNT; i++) {
+      totalWeight += this.estimateOutputWeight(i);
+    }
+    // Convert weight to vsize (rounded up)
+    return Math.ceil(totalWeight / 4);
+  }
+
+  private estimateInputWeight(inputIndex: number): number {
```

Comment:
Actually initially I started with creating a utility fn that does that using the existing caravan/bitcoin functionality .... 
You can find the code in utils.ts ....

```js

/**
 * Estimate the virtual size of a transaction based on its address type and input/output count.
 * @param addressType The type of address (P2SH, P2SH_P2WSH, or P2WSH)
 * @param inputCount Number of inputs
 * @param outputCount Number of outputs
 * @param m Number of required signers
 * @param n Total number of signers
 * @returns Estimated virtual size in vbytes
 */
export function estimateVirtualSize(
  addressType: string,
  inputCount: number,
  outputCount: number,
  m: number,
  n: number,
): number {
  const config = {
    numInputs: inputCount,
    numOutputs: outputCount,
    m,
    n,
  };

  switch (addressType) {
    case P2SH:
      return estimateMultisigP2SHTransactionVSize(config);
    case P2SH_P2WSH:
      return estimateMultisigP2SH_P2WSHTransactionVSize(config);
    case P2WSH:
      return estimateMultisigP2WSHTransactionVSize(config);
    default:
      throw new Error("Unsupported address type");
  }
}
```

But then as in PSBT  calculating the virtual size (vsize) of a transaction involves considering each input and output separately rather than relying on a single address type. So that's why implemented my own ```estimateVsize``` fn , can you please check this thoroughly to check if it looks correct to you :)

Reply:
Now this stands depreciated as we shifted from having RBF/CPFP class to the new tx anlyzer / btcTemplate approach

New commit [here](https://github.com/caravan-bitcoin/caravan/pull/114/commits/8ca1b750985cffb2d3fa3770805af53da10fa624)